### PR TITLE
Enhance support for Windows systems

### PR DESCRIPTION
**Pull request by jcarpent**
This support is related to the integration of CppAD inside Conda (see https://github.com/conda-forge/staged-recipes/pull/11300 for further details).

What I mostly did in this PR is to make CppAD a shared library on Windows systems, similarly to what it is done for Unix systems.

### CODE REVIEW

**Comment by bradbell** [comment]
I am working on putting AD in a covid-19 estimation python program for work. I will try to find time to load you pull request in a windows machine and see what it going on the the AppVeyor build.

File: test_more/general/CMakeLists.txt [hunk 1]
```diff
@@ -152,14 +152,14 @@ set_compile_flags( test_more_general "${cppad_debug_which}" "${source_list}" )
 ADD_EXECUTABLE(test_more_general EXCLUDE_FROM_ALL ${source_list})
 #
 # must first build cppad_lib
-STRING(COMPARE EQUAL "${cppad_lib}" "" empty_cppad_lib )
+STRING(COMPARE EQUAL "cppad_lib" "" empty_cppad_lib )
```
Comment (jcarpent) [inline test_more/general/CMakeLists.txt:155]: To be removed.

**Comment by jcarpent** [comment]
No problem for the delays and so on. I hope things are going well on your side.

Concerning the current on Windows, it seems that the errors are coming from https://github.com/coin-or/CppAD/blob/69d069212c039e1fecc5aba0a7ed2b0b331fe047/include/cppad/utility/thread_alloc.hpp#L197, where the static variable is not the same across the various compilations units (exe and dll). One thing to do would move some part of the implementation of the `class thread_alloc` to the lib, which will handle the global variables. @bradbell What do you think?

**Comment by bradbell** [comment]
I have started work on this pull request:

1. You are correct in that cppad_lib is always required (now) and I missed some things that should have been changed for this; e.g., changing ${cppad_lib} to cppad_lib is a good idea.

2. You have removed the CPPAD_LIB_EXPORT from the implementations of the library routines and put them in the corresponding include files. Is this because the include files appear both when compiling the library (needs an export declaration) and when compiling code that uses the library (needs an import declaration) and the export / import switch happens automatically this way ?

3. You have added the following routines to the library ?

    std::allocator<char>;
    std::basic_string<char,std::char_traits<char>,std::allocator<char> >;

4. CppAD has a define and undef convention to keep track of preprocessor macors; see include/cppad/core/undef.h The convention is to undef macros when they are no longer needed. If I execute bin/check_define.sh on your branch I get the following output:

   Check # define versus # undef names and check for addon names
    -----------------------------------------------------------------------
    27d26
    < CPPAD_EXPLICIT_TEMPLATE_IMPLEMENTATION
    check_define.sh: Error: defines and undefs do not match
    Check # define versus # undef names and check for addon names

**Comment by bradbell** [comment]
I had some questions above that I wanted to know about before merging this into the master. I tried to contact you directly by e-mail, but your address seems to have changed.

**Comment by bradbell** [comment]
It appears this pull request is still outstanding. I have having trouble understanding in the context of the current version of CppAD. Would you please describe to me what is still missing (from the current master).

**Comment by jcarpent** [comment]
Hi @bradbell, I think this feature is no more required. I will open a new PR if really needed.

**Pull request closed without merging**